### PR TITLE
Fix bugs in std::string

### PR DESCRIPTION
**Pull request by miscco**
This pulls out the bugfixes for std::string from #1735

* Missing capacity for null terminator
* Leaking memory in case of unequal allocators
* Not swapping the proxy during swap
* swapping data in case of self swap

I did keep the `_Is_element_ptr` changes, please bear with me here

### CODE REVIEW

**Review by CaseyCarter — DISMISSED** [review]

File: stl/inc/xstring
Comment (miscco) [reply stl/inc/xstring:295]: Now that I think of it I am not really sure whether we should instead simply do a `construct_at` at the position

File: tests/std/tests/P0980R1_constexpr_strings/test.cpp
Comment (miscco) [reply tests/std/tests/P0980R1_constexpr_strings/test.cpp:216]: All hail the 🛸

File: stl/inc/xstring
Comment (miscco) [reply stl/inc/xstring:295]: I believe the issue is that `_Primary_char_traits::assign` does not implement the `wmemset` optimization

Comment (CaseyCarter) [reply stl/inc/xstring:295]: Oh, derp - we can't pull in only the overload we want with `using _Primary_char_traits::assign`.

File: tests/std/tests/P0980R1_constexpr_strings/test.cpp
Comment (mnatsuhara) [inline tests/std/tests/P0980R1_constexpr_strings/test.cpp:2142]: Is there a reason that the `#ifndef __EDG__ // TRANSITION, VSO-1273296` is being moved out of individual functions?  Is it just for ease/readability?  
I originally applied it with finer granularity around specific lines of code so we could have as much EDG coverage as possible for `constexpr string`.

**Review by mnatsuhara — COMMENTED** [review]
Reviewed all changes and they look good!  I'm really happy with the additional test coverage -- thanks for going through and adding all of that :) Just one question regarding test coverage for EDG.

File: tests/std/tests/P0980R1_constexpr_strings/test.cpp
Comment (miscco) [reply tests/std/tests/P0980R1_constexpr_strings/test.cpp:2142]: If I remember correctly, then @CaseyCarter was asking about some EDG special case I applied and it turned out to be fully disabled for EDG

Given that currently it is only citing a single bug I found it clearer to move the guard to the top

Comment (mnatsuhara) [reply tests/std/tests/P0980R1_constexpr_strings/test.cpp:2142]: Do you happen to recall where that conversation took place?  Looking through the activity in #1735 and here, I don't see anything to that effect.  I searched around in Discord a bit as well, but wasn't able to find this conversation.  

If possible, I would prefer to maintain a finer granularity for this switch (only disabling specific offending lines of code) so we have as much EDG test coverage as possible.  Given how many EDG bugs we encountered in `constexpr` containers, I'd like to continue to validate what we can so the fixes we _do_ have don't regress if the EDG team is working on fixes for the remaining issues.  

I'll approve the changes as this was my only concern, and perhaps we can leave this conversation "unresolved" and have the second reviewer chime in as well.

**Review by mnatsuhara — APPROVED** [review]

File: tests/std/tests/P0980R1_constexpr_strings/test.cpp
Comment (StephanTLavavej) [reply tests/std/tests/P0980R1_constexpr_strings/test.cpp:2142]: I'll push changes to make the EDG workarounds fine-grained.

**Comment by StephanTLavavej** [comment]
I've pushed the following changes:

* Merge `main`, conflict-free.
* Fine-grained EDG workarounds, as requested by @mnatsuhara. Some are debug-only.
  + These workarounds were *almost* verbose enough to abandon the attempt, but Miya made an excellent case for preventing regressions in brittle compiler codepaths and that's worth the effort.
  + Also, this replaces the "MSVC internal testing" guard with an EDG guard. The affected code works (or does a good job of appearing to work) with both the released VS Preview and internal builds of MSVC, but doesn't work for EDG.
  + As an aside, the "MSVC internal testing" guard was recently renamed to add an underscore: `_MSVC_INTERNAL_TESTING`
* After GH-2064 was merged, we guard `is_constant_evaluated` with `_HAS_CXX20`.
* Cleanup: use `_My_data`/`_Right_data` to avoid verbosity. Now that `_My_large`/`_Right_large` aren't needed in multiple places, "inline" them with no loss of clarity.
* The `_Is_elem_cptr_v` change was desirable, but must be reverted due to `/clr:pure` failures of the form `test.obj : fatal error LNK1179: invalid or corrupt file: duplicate COMDAT '???__E??$_Is_elem_cptr_v@[...blah...]'`. (I believe it dislikes variable templates other than at namespace scope.)

Thanks for fixing all these bugs, especially the one where swap didn't think it needed to swap proxies for the small-small case. I was *sure* that change wasn't necessary, until I attempted to revert it and discovered a test failure, and then I was *sure* the test was bogus, until I looked further and saw that the test was quite strict about the behavior of equality. Nicely done!

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks :lady_beetle: for :ant: fixing :bug: these :beetle: bugs! :honeybee:

**Pull request merged**
